dhctl/pkg/preflight/registry.go

```diff
@@ -201,3 +212,228 @@ func setupSSHTunnelToProxyAddr(sshCl *ssh.Client, proxyUrl *url.URL) (*frontend.
+	ctx, cancel := context.WithTimeout(context.Background(), httpClientTimeoutSec*time.Second)
+	defer cancel()
+
+	authData, err := pc.metaConfig.Registry.Auth()
+	if err != nil {
+		return err
+	}
+
+	if authData == "" {
+		return fmt.Errorf("%w, credentials are not specified. If you are using CE edition in a closed environment, this check can be skipped by specifying the --preflight-skip-registry-credential flag", ErrAuthFailed)
+	}
+
+	return checkRegistryAuth(ctx, pc.metaConfig, authData)
+}
+
+func prepareRegistryRequest(ctx context.Context, metaConfig *config.MetaConfig, authData string) (*http.Request, error) {
+	registryURL := &url.URL{Scheme: metaConfig.Registry.Scheme, Host: metaConfig.Registry.Address, Path: registryPath}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, registryURL.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("prepare registry request: %w", err)
+	}
+	if authData != "" {
+		req.Header.Add("Authorization", "Basic "+authData)
+	}
+
+	return req, nil
+}
+
+func prepareAuthRequest(
+	ctx context.Context,
+	authURL string,
+	registryService string,
+	authData string,
+	metaConfig *config.MetaConfig,
+) (*http.Request, error) {
+	authURLValues := url.Values{}
+	authURLValues.Add("service", registryService)
+	authURLValues.Add("scope", fmt.Sprintf("repository:%s:pull", strings.TrimLeft(metaConfig.Registry.Path, "/")))
+
+	authURL = fmt.Sprintf("%s?%s", authURL, authURLValues.Encode())
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, authURL, nil)
+	if err != nil {
+		return nil, fmt.Errorf("prepare auth request: %w", err)
+	}
+
+	req.Header.Add("Authorization", "Basic "+authData)
+
+	return req, nil
+}
+
+func prepareAuthHTTPClient(metaConfig *config.MetaConfig) (*http.Client, error) {
+	client := &http.Client{}
+	httpTransport := http.DefaultTransport.(*http.Transport).Clone()
+
+	if strings.ToLower(metaConfig.Registry.Scheme) == "http" {
+		httpTransport.TLSClientConfig = &tls.Config{
+			InsecureSkipVerify: true,
+		}
+	}
+
+	if len(metaConfig.Registry.CA) == 0 {
+		client.Transport = httpTransport
+		return client, nil
+	}
+
+	certPool := x509.NewCertPool()
+	if ok := certPool.AppendCertsFromPEM([]byte(metaConfig.Registry.CA)); !ok {
+		return nil, fmt.Errorf("invalid cert in CA PEM")
+	}
+
+	httpTransport.TLSClientConfig = &tls.Config{
+		RootCAs: certPool,
+	}
+
+	client.Transport = httpTransport
+
+	return client, nil
+}
+
+func getAuthRealmAndService(ctx context.Context, metaConfig *config.MetaConfig, client *http.Client) (string, string, error) {
+	authURL := ""
+	registryService := ""
+
+	req, err := prepareRegistryRequest(ctx, metaConfig, "")
+	if err != nil {
+		return authURL, registryService, err
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return authURL, registryService, fmt.Errorf("cannot auth in registry. %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.Header.Get("Docker-Distribution-API-Version") != "registry/2.0" {
```

Comment:
Does it work for all registries as Nexus or Quay?

Reply:
It is a part of registry v2 api spec